### PR TITLE
make the MPI communciator setable

### PR DESCRIPTION
**Pull request by burlen**
start to work on #183 #201 #202 

when this is ready for review I will remove the WIP from the pull request name

### CODE REVIEW

**Comment by burlen** [comment]
gcc in Ubuntu 14.04 is segfaulting when compiling.
http://cdash.hpcvis.com/viewBuildError.php?buildid=4091

**Comment by taobrienlbl** [comment]
@burlen: [CDASH](http://cdash.hpcvis.com/testDetails.php?test=53421&build=4145) shows a test failure for test_binary_segmentation on Ubuntu_14.04_LTS_user_specified_communicator_941954d: `Attempting to use an MPI routine before initializing MPICH`

**Comment by burlen** [comment]
@taobrienlbl I think this is ready for testing and review, would you please test it and see if it resolves your issue?

File: core/teca_algorithm.cxx [hunk 1]
```diff
@@ -121,14 +127,50 @@ teca_algorithm_internals::teca_algorithm_internals()
     name("teca_algorithm"),
     data_cache_size(1),
     modified(1),
-    exec(teca_algorithm_executive::New())
+    exec(teca_algorithm_executive::New()),
+#if defined(TECA_HAS_MPI)
```
Comment (taobrienlbl) [inline core/teca_algorithm.cxx]: @burlen - could you please clarify what this block is doing (perhaps clarifying w/ comments in the code)?

Is this setting a default for the communicator?

**Comment by burlen** [comment]
`test_binary_segmentation` is failing on Travis Ubuntu 14.04,
```
Attempting to use an MPI routine before initializing MPICH
```
http://cdash.hpcvis.com/testDetails.php?test=53421&build=4146

This is a serial test and as far as I can tell makes no MPI calls. I have tested locally with Fedora 28 and both OpenMPI and MPICH and am not able to reproduce. I have also made an effort to look at all the code paths through the test and am not seeing anywhere that would use MPI without first checking to see if it was initialized. I could be wrong but this is looking a lot like a bug in the very likely ancient MPICH deployed on Ubuntu 14.04.

**Comment by taobrienlbl** [comment]
@burlen , in addition to the reference to `MPI_COMM_WORLD` that I just commented on for `core/teca_algorithm.cxx`, there are two other places where `MPI_COMM_WORLD` is hard-coded in core TECA code: in `teca_parallel_id.cxx` and `teca_mpi_manager.cxx`.  

```
./core/teca_parallel_id.cxx:22:        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
./core/teca_mpi_manager.cxx:24:    MPI_Comm_rank(MPI_COMM_WORLD, &m_rank);
./core/teca_mpi_manager.cxx:25:    MPI_Comm_size(MPI_COMM_WORLD, &m_size);
```

This might be intentional and necessary, but if so, it isn't clear from looking at the code why.  If it is intentional/necessary, could you please add comments into the code to clarify?

File: core/teca_algorithm.cxx (hunk 1 above)
Comment (burlen) [reply core/teca_algorithm.cxx]: yes, that is where the default communicator is set. If no one calls `set_communicator` then then `MPI_COMM_WORLD` is returned from `get_communicator`. I'll add a comment.

**Comment by burlen** [comment]
> @burlen , in addition to the reference to `MPI_COMM_WORLD` that I just commented on for `core/teca_algorithm.cxx`, there are two other places where `MPI_COMM_WORLD` is hard-coded in core TECA code: in `teca_parallel_id.cxx` and `teca_mpi_manager.cxx`.
> 
> ```
> ./core/teca_parallel_id.cxx:22:        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
> ./core/teca_mpi_manager.cxx:24:    MPI_Comm_rank(MPI_COMM_WORLD, &m_rank);
> ./core/teca_mpi_manager.cxx:25:    MPI_Comm_size(MPI_COMM_WORLD, &m_size);
> ```
> 
> This might be intentional and necessary, but if so, it isn't clear from looking at the code why. If it is intentional/necessary, could you please add comments into the code to clarify?

`teca_parallel_id` should always use the rank in `MPI_COMM_WORLD`. `teca_mpi_manager` is a helper that ensures MPI finalize is called in tests and applications, when errors occur, and I can't think of a use case it would make sense for it not to use `MPI_COMM_WORLD`.

I can add a comment

**Comment by burlen** [comment]
> @burlen: [CDASH](http://cdash.hpcvis.com/testDetails.php?test=53421&build=4145) shows a test failure for test_binary_segmentation on Ubuntu_14.04_LTS_user_specified_communicator_941954d: `Attempting to use an MPI routine before initializing MPICH`

I added a comment about this above. Long story short, as far as I can tell it is a bug in the version of MPICH deployed with  Ubuntu 14.04. I welcome your fresh eyes on it though. Could you take a look and see if I missing something stupid?

**Comment by taobrienlbl** [comment]
@burlen - thanks for clarifying, and comments will be perfect.

Regarding the build failure, would it make sense to see if the test replicates on cori, since that's one of our current target platforms?  The other thing we could consider is that @elbashandy could create a 14.04 docker image in which we could use debugging tools to try to diagnose where in the code this is being triggered and whether it for sure is an MPICH bug.

If you're confident its a bug w/o doing further testing, I'm OK with proceeding; though I'm leaning toward being cautious, since this could also indicate a subtle bug somewhere in the TECA code that could be hard to track down if it pops up later on in an application.

**Comment by burlen** [comment]
> @burlen - thanks for clarifying, and comments will be perfect.
> 
> Regarding the build failure, would it make sense to see if the test replicates on cori, since that's one of our current target platforms? The other thing we could consider is that @elbashandy could create a 14.04 docker image in which we could use debugging tools to try to diagnose where in the code this is being triggered and whether it for sure is an MPICH bug.
> 
> If you're confident its a bug w/o doing further testing, I'm OK with proceeding; though I'm leaning toward being cautious, since this could also indicate a subtle bug somewhere in the TECA code that could be hard to track down if it pops up later on in an application.

My 2 cents is that it is not worth messing any more with 14.04. Abdel's new 18.04 docker based environment should replace 14.04 asap.

I agree that we should test on Cori prior to the release. perhaps we even set up a nightly dashboard submission from cori, so that we are continuously testing on cori.

File: core/teca_algorithm.h [hunk 2]
```diff
@@ -28,6 +29,17 @@ class teca_algorithm : public std::enable_shared_from_this<teca_algorithm>
...
 
+    // set/get the communicator to use at this stage of the pipeline this has
+    // no influence on other stages. We duplicate the passed in communicator
+    // providing an isolated communication space for subsequent operations. By
+    // default the communicator is initialized to MPI_COMM_WORLD, here it is not
+    // duplicated. Thus to put an algorithm into a unique communication space
+    // one should explicitly set a communicator. When an algorithm should not
+    // use MPI, for instance when it is in a nested pipeline, one may set the
+    // communicator to MPI_COMM_SELF.
+    void set_communicator(MPI_Comm comm);
```
Comment (burlen) [inline core/teca_algorithm.h]: @taobrienlbl comments added

**Comment by taobrienlbl** [comment]
Attempting to re-open the PR, following comments in #183

**Pull request merged**